docs/root/intro/arch_overview/matching/matching_api.rst

```diff
@@ -0,0 +1,74 @@
+.. _arch_overview_matching_api:
+
+Matching API
+============
+
+Envoy makes use of a matching API to allow the various subsystems to express actions that should
+be performed based on incoming data.
+
+The matching API is designed as a tree structure to allow for sublinear matching algorithms for
+performance better than the linear list matching as seen in Envoy's HTTP routing. It makes heavy
+use of extension points to make it easy to extend to different inputs based on protocol or
+environment data as well as custom sublinear matchers and direct matchers.
+
+Within supported environments (currently only HTTP filters), a wrapper proto can be used to
+instantiate a matching filter associated with the wrapped structure:
+
+.. literalinclude:: _include/simple.yaml
+    :language: yaml
+
+The above example wraps a HTTP filter (the
+:ref:`HTTPFault <envoy_v3_api_msg_extensions.filters.http.fault.v3.HttpFault>` filter) in an
+:ref:`ExtensionWithMatcher <envoy_v3_api_msg_extensions.common.matching.v3.ExtensionWithMatcher>`,
+allowing us to define a match tree to be evaluated in conjunction with evaluation of the wrapped
+filter. Prior to data being made available to the filter, it will be provided to the match tree,
+which will then attempt to evaluate the matching rules with the provided data, triggering an
+action if match evaluation completes in an action.
+
+In the above example, we are specifying that we want to match on the incoming request header
+`some-header` by setting the `input` to
+:ref:`HttpRequestHeaderMatchInput <envoy_v3_api_msg_type.matcher.v3.HttpRequestHeaderMatchInput>`
+and configuring the header key to use. Using the value contained by this header, the provided
+`exact_match_map` specifies which values we care about: we've configured a single value
+(`some_value_to_match_on`) to match against. As a result, this config means that if we
+receive a request which contains `some-header: some_value_to_match_on` as a header, the
+:ref:`SkipFilter <envoy_v3_api_msg_extensions.filters.common.matcher.action.v3.SkipFilter>`
+action will be resolved (causing the associated HTTP filter to be skipped). If no such header is
+present, no action will be resolved and the filter will be applied as usual.
+
+.. literalinclude:: _include/complicated.yaml
+    :language: yaml
+
+Above is a slightly more complicated example which combines a top level tree matcher with a
+linear matcher. While the tree matchers provide very efficient matching, they are not very
+expressive. The list matcher can be used to provide a much richer matching API, and can be combined
+with the tree matcher in an arbitrary order. The example describes the following match logic: skip
+the filter if `some-header: skip_filter` is present and `second-header` is set to *either* `foo` or
+`bar`.
+
+HTTP Filter Iteration Impact
```

Comment:
Is it worth putting an attention block in here with a link to the open issue tracking match schema checking or similar?